File: src/main/cpp/subsystems/Drivetrain.cpp
#include "subsystems/Drivetrain.h"
#include <frc/smartdashboard/SmartDashboard.h>

Drivetrain::Drivetrain()
{

}

void Drivetrain::UpdateTelemetry()
{

}

void Drivetrain::RegisterDataEntries(wpi::log::DataLog &log)
{

}

void Drivetrain::LogDataEntries(wpi::log::DataLog &log)
{

}

void Drivetrain::ConfigureSystem()
{
  m_imu.ZeroGyroBiasNow(50);
  ResetOdometry(GetPose());

  m_yawLockPID.EnableContinuousInput(-0.5_tr / 1_rad, 0.5_tr / 1_rad);
  m_yawLockPID.SetTolerance(3_deg / 1_rad);

  // Display Robot position on field
  frc::SmartDashboard::PutData("Field", &m_fieldDisplay);

  m_frontLeft.ConfigureSystem();
  m_frontRight.ConfigureSystem();
  m_backLeft.ConfigureSystem();
  m_backRight.ConfigureSystem();
}

void Drivetrain::Drive(pathplanner::PathPlannerTrajectory::PathPlannerState targetState)
{
  auto command = m_ppController.calculate(
      m_poseEstimator.GetEstimatedPosition(),
      targetState);

  command.omega = -1 *command.omega;

  Drive(command, false);
}

void Drivetrain::Drive(frc::ChassisSpeeds speed,
                       bool openLoop)
{

  // Heading Lock
  // constexpr auto noRotThreshold = 0.1_deg_per_s;
  // m_YawLockActive = units::math::abs(rot) < noRotThreshold;

  // const auto trans_mag = units::math::sqrt(xSpeed * xSpeed + ySpeed * ySpeed);
  
  // if (m_YawLockActive && trans_mag > 0.1_mps)
  // {
  //   auto r = m_yawLockPID.Calculate(GetYaw().Radians() / 1_rad);
  //   if (!m_yawLockPID.AtSetpoint())
  //   {
  //     // Robot will automatically maintain current yaw
  //     rot = units::radians_per_second_t{-r};
  //   } else {
  //     m_yawLockPID.Reset();
  //   }
  // }
  // else
  // {
  //   // Manual control, save the current yaw.
  //   m_yawLockPID.Reset();
  //   m_yawLockPID.SetSetpoint(GetYaw().Radians() / 1_rad);
  // }

  m_input = speed;
  
  // Transform Field Oriented command to a Robot Relative Command
  if (m_fieldRelative)
  {
    m_command = frc::ChassisSpeeds::FromFieldRelativeSpeeds(speed.vx, speed.vy, speed.omega, GetYaw());
  }
  else
  {
    m_command = speed;
  }

  m_originalCommand = m_command;

  // Calculate desired swerve states
  auto states = m_kinematics.ToSwerveModuleStates(m_command);
  // m_kinematics.DesaturateWheelSpeeds(&states, m_command, 16_fps, 16_fps, 360_deg_per_s);
  m_kinematics.DesaturateWheelSpeeds(&states, m_command, kTrueMaxSpeedLinear, kTrueMaxSpeedLinear, kTrueMaxSpeedAngular);

  m_command = m_kinematics.ToChassisSpeeds(states);

  // Set State of Each Module
  auto [fl, fr, bl, br] = states;
  m_frontLeft.SetModule(fl, openLoop);
  m_frontRight.SetModule(fr, openLoop);
  m_backLeft.SetModule(bl, openLoop);
  m_backRight.SetModule(br, openLoop);
}

void Drivetrain::Stop()
{
  m_frontLeft.Stop();
  m_frontRight.Stop();
  m_backLeft.Stop();
  m_backRight.Stop();
}

frc::Rotation2d Drivetrain::GetIMUYaw()
{
  return m_imu.GetRotation2d();
}

frc::Rotation2d Drivetrain::GetYaw()
{
  return GetPose().Rotation();
}

void Drivetrain::UpdateOdometry()
{
  auto estPose = m_poseEstimator.Update(GetIMUYaw(),{
                         m_frontLeft.GetModulePosition(),
                         m_frontRight.GetModulePosition(),
                         m_backLeft.GetModulePosition(),
                         m_backRight.GetModulePosition()});
  
  m_robotVelocity = m_kinematics.ToChassisSpeeds({m_frontLeft.GetState(),
                                                  m_frontRight.GetState(),
                                                  m_backLeft.GetState(),
                                                  m_backRight.GetState()});

  m_estimatedPose->SetPose(estPose);
}

void Drivetrain::ResetYaw(const frc::Rotation2d &yaw)
{
  m_imu.SetYaw(yaw.Radians() / 1_deg, 50);
  m_yawLockPID.Reset();
  m_yawLockPID.SetSetpoint(yaw.Radians() / 1_rad);
}

void Drivetrain::ResetOdometry(const frc::Pose2d &pose)
{
  m_poseEstimator.ResetPosition(GetIMUYaw(), {
                         m_frontLeft.GetModulePosition(),
                         m_frontRight.GetModulePosition(),
                         m_backLeft.GetModulePosition(),
                         m_backRight.GetModulePosition()}, pose);
}

frc::ChassisSpeeds Drivetrain::GetChassisSpeeds()
{
  return m_robotVelocity;
}

void Drivetrain::InitSendable(wpi::SendableBuilder &builder)
{
  builder.SetSmartDashboardType("DriveBase");
  builder.SetActuator(true);

  // Modules
  m_frontLeft.InitSendable(builder);
  m_frontRight.InitSendable(builder);
  m_backLeft.InitSendable(builder);
  m_backRight.InitSendable(builder);

  builder.AddDoubleProperty("gyro", [this] { return units::degree_t{m_imu.GetYaw()} / 1_rad; }, nullptr);
  
  // Pose
  builder.AddDoubleProperty(
      "poseEstimator/x", [this] { return m_poseEstimator.GetEstimatedPosition().X().value(); }, nullptr);
  builder.AddDoubleProperty(
      "poseEstimator/y", [this] { return m_poseEstimator.GetEstimatedPosition().Y().value(); }, nullptr);
  builder.AddDoubleProperty(
      "poseEstimator/yaw", [this] { return m_poseEstimator.GetEstimatedPosition().Rotation().Degrees().value(); }, nullptr);

  // builder.AddDoubleProperty(
  //     "odometry/x", [this] { return m_odometry.GetPose().X().value(); }, nullptr);
  // builder.AddDoubleProperty(
  //     "odometry/y", [this] { return m_odometry.GetPose().Y().value(); }, nullptr);
  // builder.AddDoubleProperty(
  //     "odometry/yaw", [this] { return m_odometry.GetPose().Rotation().Degrees().value(); }, nullptr);

  // Command
    builder.AddDoubleProperty(
      "input/x", [this] { return m_input.vx / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
      "input/y", [this] { return m_input.vy / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
      "input/yaw", [this] { return m_input.omega / 1_rad_per_s; }, nullptr);

  builder.AddDoubleProperty(
      "cmd/x", [this] { return m_command.vx / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
      "cmd/y", [this] { return m_command.vy / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
      "cmd/yaw", [this] { return m_command.omega / 1_rad_per_s; }, nullptr);

  builder.AddDoubleProperty(
      "demand/x", [this] { return m_originalCommand.vx / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
      "demand/y", [this] { return m_originalCommand.vy / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
      "demand/yaw", [this] { return m_originalCommand.omega / 1_rad_per_s; }, nullptr);

  // State
    builder.AddDoubleProperty(
      "state/x", [this] { return m_robotVelocity.vx / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
      "state/y", [this] { return m_robotVelocity.vy / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
      "state/yaw", [this] { return m_robotVelocity.omega / 1_rad_per_s; }, nullptr);

  // Operating Mode
  builder.AddBooleanProperty(
      "cmd/fieldRelative", [this] { return m_fieldRelative; }, nullptr);

  builder.AddDoubleProperty(
      "yawpid/setpoint", [this] { return m_yawLockPID.GetSetpoint(); }, nullptr); 
  builder.AddDoubleProperty(
      "yawpid/error", [this] { return m_yawLockPID.GetPositionError(); }, nullptr);
  builder.AddBooleanProperty(
      "yawpid/atSetpoint", [this] { return m_yawLockPID.AtSetpoint(); }, nullptr);
}

units::ampere_t Drivetrain::SimPeriodic(units::volt_t battery)
{

  // Simulated IMU
  m_imuSimCollection.AddHeading(GetChassisSpeeds().omega * 20_ms / 1_deg);

  return m_frontLeft.SimPeriodic(battery) +
    m_frontRight.SimPeriodic(battery) +
    m_backLeft.SimPeriodic(battery) +
    m_backRight.SimPeriodic(battery);
}

void Drivetrain::SimInit()
{
  m_frontLeft.SimInit();
  m_frontRight.SimInit();
  m_backLeft.SimInit();
  m_backRight.SimInit();
}

ErrorCode Drivetrain::SeedEncoders() {
  auto error = m_frontLeft.SeedTurnMotor();
  if (error != ErrorCode::OK) {
    return error;
  }

  error = m_frontRight.SeedTurnMotor();
  if (error != ErrorCode::OK) {
    return error;
  }

  error = m_backLeft.SeedTurnMotor();
  if (error != ErrorCode::OK) {
    return error;
  }

  return m_backRight.SeedTurnMotor();
}

bool Drivetrain::Active()
{
  return units::math::abs(m_frontLeft.AngularVelocity()) > 0.01_rad_per_s
    || units::math::abs(m_frontRight.AngularVelocity()) > 0.01_rad_per_s
    || units::math::abs(m_backLeft.AngularVelocity()) > 0.01_rad_per_s
    || units::math::abs(m_backRight.AngularVelocity()) > 0.01_rad_per_s;
}

frc::Pose2d Drivetrain::GetPose()
{
  return m_poseEstimator.GetEstimatedPosition();
}

bool Drivetrain::GetFieldCentric()
{
  return m_fieldRelative;
}

void Drivetrain::SetFieldCentric(bool fieldRelative)
{
  m_fieldRelative = fieldRelative;
}

File: src/main/java/io/robojackets/subsystems/Drivetrain.java
package io.robojackets.subsystems;

import com.ctre.phoenix.ErrorCode;
import com.ctre.phoenix.sensors.BasePigeonSimCollection;
import com.ctre.phoenix.sensors.WPI_Pigeon2;
import edu.wpi.first.math.controller.HolonomicDriveController;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.ProfiledPIDController;
import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.math.trajectory.Trajectory.State;
import edu.wpi.first.math.trajectory.TrapezoidProfile;
import edu.wpi.first.util.datalog.DataLog;
import edu.wpi.first.util.sendable.SendableBuilder;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.FieldObject2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import io.robojackets.config.FeedForwardConfig;
import io.robojackets.config.PIDConfig;
import io.robojackets.config.SwerveModuleConfig;
import io.robojackets.lib.UnitConversion;

public class Drivetrain extends Subsystem {
  public static final double kMaxSpeedLinearMetersPerSecond = 16 * UnitConversion.METERS_PER_FOOT;
  public static final double kMaxSpeedAngularRadiansPerSecond =
      360 / UnitConversion.DEGREES_PER_RADIAN;
  public static final double kMaxAccelerationLinearMetersPerSecondPerSecond =
      20 * UnitConversion.METERS_PER_FOOT;
  public static final double kWheelDistance = 20.5 * UnitConversion.METERS_PER_INCH;

  private boolean fieldRelative = true;
  Translation2d frontLeftLocation = new Translation2d(kWheelDistance / 2, kWheelDistance / 2);
  Translation2d frontRightLocation = new Translation2d(kWheelDistance / 2, -kWheelDistance / 2);
  Translation2d backLeftLocation = new Translation2d(-kWheelDistance / 2, kWheelDistance / 2);
  Translation2d backRightLocation = new Translation2d(-kWheelDistance / 2, -kWheelDistance / 2);

  WPI_Pigeon2 imu = new WPI_Pigeon2(30);
  BasePigeonSimCollection imuSim = imu.getSimCollection();

  Field2d fieldDisplay = new Field2d();
  FieldObject2d estimatedPose = fieldDisplay.getRobotObject();

  ChassisSpeeds command = new ChassisSpeeds();
  ChassisSpeeds originalCommand = new ChassisSpeeds();
  ChassisSpeeds measuredVelocity = new ChassisSpeeds();

  boolean yawLockActive = true;
  PIDController yawLockPID = new PIDController(5, 0, 0);

  SwerveDriveKinematics kinematics =
      new SwerveDriveKinematics(
          frontLeftLocation, frontRightLocation, backLeftLocation, backRightLocation);

  SwerveModuleConfig frontLeftConfig =
      SwerveModuleConfig.builder()
          .angleOffsetRadians(-128.496 / UnitConversion.DEGREES_PER_RADIAN)
          .drivePIDConfig(PIDConfig.builder().kP(0.0092534).build())
          .turnPIDConfig(PIDConfig.builder().kP(0.098996).kD(0.00055669).build())
          .driveFeedForwardConfig(
              FeedForwardConfig.builder().kS(0.66323).kV(2.1798).kA(0.15467).build())
          .turnFeedForwardConfig(
              FeedForwardConfig.builder().kS(0.72584).kV(0.21377).kA(0.0027946).build())
          .maxTurnVelocityRadiansPerSecond(36.0)
          .maxTurnAccelRadiansPerSecondPerSecond(1000.0)
          .build();

  SwerveModuleConfig frontRightConfig =
      SwerveModuleConfig.builder()
          .angleOffsetRadians(78.838 / UnitConversion.DEGREES_PER_RADIAN)
          .drivePIDConfig(PIDConfig.builder().kP(0.0092534).build())
          .turnPIDConfig(PIDConfig.builder().kP(0.098996).kD(0.00055669).build())
          .driveFeedForwardConfig(
              FeedForwardConfig.builder().kS(0.66323).kV(2.1798).kA(0.15467).build())
          .turnFeedForwardConfig(
              FeedForwardConfig.builder().kS(0.72584).kV(0.21377).kA(0.0027946).build())
          .maxTurnVelocityRadiansPerSecond(36.0)
          .maxTurnAccelRadiansPerSecondPerSecond(1000.0)
          .build();

  SwerveModuleConfig backLeftConfig =
      SwerveModuleConfig.builder()
          .angleOffsetRadians(-4.219 / UnitConversion.DEGREES_PER_RADIAN)
          .drivePIDConfig(PIDConfig.builder().kP(0.0092534).build())
          .turnPIDConfig(PIDConfig.builder().kP(0.098996).kD(0.00055669).build())
          .driveFeedForwardConfig(
              FeedForwardConfig.builder().kS(0.66323).kV(2.1798).kA(0.15467).build())
          .turnFeedForwardConfig(
              FeedForwardConfig.builder().kS(0.72584).kV(0.21377).kA(0.0027946).build())
          .maxTurnVelocityRadiansPerSecond(36.0)
          .maxTurnAccelRadiansPerSecondPerSecond(1000.0)
          .build();

  SwerveModuleConfig backRightConfig =
      SwerveModuleConfig.builder()
          .angleOffsetRadians(-60.645 / UnitConversion.DEGREES_PER_RADIAN)
          .drivePIDConfig(PIDConfig.builder().kP(0.0092534).build())
          .turnPIDConfig(PIDConfig.builder().kP(0.098996).kD(0.00055669).build())
          .driveFeedForwardConfig(
              FeedForwardConfig.builder().kS(0.66323).kV(2.1798).kA(0.15467).build())
          .turnFeedForwardConfig(
              FeedForwardConfig.builder().kS(0.72584).kV(0.21377).kA(0.0027946).build())
          .maxTurnVelocityRadiansPerSecond(36.0)
          .maxTurnAccelRadiansPerSecondPerSecond(1000.0)
          .build();

  SwerveModule frontLeft = new SwerveModule("FL", 0, 1, 20, frontLeftConfig);
  SwerveModule frontRight = new SwerveModule("FR", 2, 3, 21, frontRightConfig);
  SwerveModule backLeft = new SwerveModule("BL", 4, 5, 22, backLeftConfig);
  SwerveModule backRight = new SwerveModule("BR", 6, 7, 23, backRightConfig);

  SwerveDrivePoseEstimator poseEstimator =
      new SwerveDrivePoseEstimator(
          kinematics,
          GetYaw(),
          new SwerveModulePosition[] {
            frontLeft.GetModulePosition(),
            frontRight.GetModulePosition(),
            backLeft.GetModulePosition(),
            backRight.GetModulePosition()
          },
          new Pose2d());

  HolonomicDriveController trajectoryController =
      new HolonomicDriveController(
          new PIDController(2.0, 0.0, 0.0),
          new PIDController(2.0, 0.0, 0.0),
          new ProfiledPIDController(
              2.0,
              0.0,
              0.0,
              new TrapezoidProfile.Constraints(
                  360 / UnitConversion.DEGREES_PER_RADIAN,
                  720 / UnitConversion.DEGREES_PER_RADIAN)));

  public Rotation2d GetYaw() {
    return Rotation2d.fromDegrees(imu.getYaw());
  }

  @Override
  public void initSendable(SendableBuilder builder) {
    builder.setSmartDashboardType("DriveBase");
    builder.setActuator(true);

    // Modules
    frontLeft.initSendable(builder);
    frontRight.initSendable(builder);
    backLeft.initSendable(builder);
    backRight.initSendable(builder);

    builder.addDoubleProperty("gyro", () -> GetYaw().getRadians(), null);

    // Pose
    builder.addDoubleProperty(
        "poseEstimator/x", () -> poseEstimator.getEstimatedPosition().getX(), null);
    builder.addDoubleProperty(
        "poseEstimator/y", () -> poseEstimator.getEstimatedPosition().getY(), null);
    builder.addDoubleProperty(
        "poseEstimator/yaw",
        () -> poseEstimator.getEstimatedPosition().getRotation().getRadians(),
        null);

    // Command
    builder.addDoubleProperty("cmd/x", () -> command.vxMetersPerSecond, null);
    builder.addDoubleProperty("cmd/y", () -> command.vyMetersPerSecond, null);
    builder.addDoubleProperty("cmd/yaw", () -> command.omegaRadiansPerSecond, null);
    builder.addDoubleProperty("demand/x", () -> originalCommand.vxMetersPerSecond, null);
    builder.addDoubleProperty("demand/y", () -> originalCommand.vyMetersPerSecond, null);
    builder.addDoubleProperty("demand/yaw", () -> originalCommand.omegaRadiansPerSecond, null);

    // State
    builder.addDoubleProperty("state/x", () -> measuredVelocity.vxMetersPerSecond, null);
    builder.addDoubleProperty("state/y", () -> measuredVelocity.vyMetersPerSecond, null);
    builder.addDoubleProperty("state/yaw", () -> measuredVelocity.omegaRadiansPerSecond, null);

    // Operating Mode
    builder.addBooleanProperty("cmd/fieldRelative", () -> fieldRelative, null);
    builder.addDoubleProperty("yawpid/setpoint", () -> yawLockPID.getSetpoint(), null);
    builder.addDoubleProperty("yawpid/error", () -> yawLockPID.getPositionError(), null);
    builder.addBooleanProperty("yawpid/atSetpoint", () -> yawLockPID.atSetpoint(), null);
  }

  @Override
  public void ConfigureSystem() {
    imu.zeroGyroBiasNow(50);

    ResetYaw(new Rotation2d());

    yawLockPID.enableContinuousInput(-Math.PI, Math.PI);
    yawLockPID.setTolerance(3.0 / 180.0 * Math.PI);

    SmartDashboard.putData("Field", fieldDisplay);

    frontLeft.ConfigureSystem();
    frontRight.ConfigureSystem();
    backLeft.ConfigureSystem();
    backRight.ConfigureSystem();
  }

  public void Drive(State state, Rotation2d yaw) {
    ChassisSpeeds command =
        trajectoryController.calculate(poseEstimator.getEstimatedPosition(), state, yaw);

    Drive(
        command.vxMetersPerSecond, command.vyMetersPerSecond, command.omegaRadiansPerSecond, false);
  }

  public void Drive(
      double xSpeedMetersPerSecond,
      double ySpeedMetersPerSecond,
      double rotationRadiansPerSecond,
      boolean openLoop) {
    final double no_rotation_threshold_RadiansPerSecond = 0.1 / UnitConversion.DEGREES_PER_RADIAN;
    yawLockActive = Math.abs(rotationRadiansPerSecond) < no_rotation_threshold_RadiansPerSecond;

    double translation_magnitude_MetersPerSecond =
        Math.hypot(xSpeedMetersPerSecond, ySpeedMetersPerSecond);

    if (yawLockActive && translation_magnitude_MetersPerSecond > 0.1) {
      double rot_demand = yawLockPID.calculate(GetYaw().getRadians());
      if (!yawLockPID.atSetpoint()) {
        rotationRadiansPerSecond = rot_demand;
      } else {
        yawLockPID.reset();
      }
    } else {
      yawLockPID.reset();
      yawLockPID.setSetpoint(GetYaw().getRadians());
    }

    if (fieldRelative) {
      command =
          ChassisSpeeds.fromFieldRelativeSpeeds(
              xSpeedMetersPerSecond, ySpeedMetersPerSecond, rotationRadiansPerSecond, GetYaw());
    } else {
      command =
          new ChassisSpeeds(xSpeedMetersPerSecond, ySpeedMetersPerSecond, rotationRadiansPerSecond);
    }

    originalCommand = command;

    SwerveModuleState[] states = kinematics.toSwerveModuleStates(command);
    SwerveDriveKinematics.desaturateWheelSpeeds(states, kMaxSpeedLinearMetersPerSecond);

    command = kinematics.toChassisSpeeds(states);

    frontLeft.SetModule(states[0], openLoop);
    frontRight.SetModule(states[1], openLoop);
    backLeft.SetModule(states[2], openLoop);
    backRight.SetModule(states[3], openLoop);
  }

  public void Stop() {
    frontLeft.Stop();
    frontRight.Stop();
    backLeft.Stop();
    backRight.Stop();
  }

  public void UpdateOdometry() {
    Pose2d estPose =
        poseEstimator.update(
            GetYaw(),
            new SwerveModulePosition[] {
              frontLeft.GetModulePosition(),
              frontRight.GetModulePosition(),
              backLeft.GetModulePosition(),
              backRight.GetModulePosition()
            });

    measuredVelocity =
        kinematics.toChassisSpeeds(
            frontLeft.GetState(), frontRight.GetState(), backLeft.GetState(), backRight.GetState());

    edu.wpi.first.math.geometry.Pose2d actualPose =
        new edu.wpi.first.math.geometry.Pose2d(
            estPose.getX(),
            estPose.getY(),
            edu.wpi.first.math.geometry.Rotation2d.fromDegrees(estPose.getRotation().getDegrees()));

    estimatedPose.setPose(actualPose);
  }

  public void ResetYaw(Rotation2d heading) {
    imu.setYaw(heading.getDegrees(), 50);
    Pose2d pose = new Pose2d(GetPose().getTranslation(), heading);

    poseEstimator.resetPosition(
        heading,
        new SwerveModulePosition[] {
          frontLeft.GetModulePosition(),
          frontRight.GetModulePosition(),
          backLeft.GetModulePosition(),
          backRight.GetModulePosition()
        },
        pose);

    yawLockPID.reset();
    yawLockPID.setSetpoint(heading.getRadians());
  }

  public void ResetOdometry(Pose2d pose) {
    poseEstimator.resetPosition(
        GetYaw(),
        new SwerveModulePosition[] {
          frontLeft.GetModulePosition(),
          frontRight.GetModulePosition(),
          backLeft.GetModulePosition(),
          backRight.GetModulePosition()
        },
        pose);
    yawLockPID.reset();
    yawLockPID.setSetpoint(GetYaw().getRadians());
  }

  @Override
  public void RegisterDataEntries(DataLog log) {
    // TODO Auto-generated method stub

  }

  @Override
  public void LogDataEntries(DataLog log) {
    // TODO Auto-generated method stub

  }

  @Override
  public void SimInit() {
    frontLeft.SimInit();
    frontRight.SimInit();
    backLeft.SimInit();
    backRight.SimInit();
  }

  @Override
  public void SimPeriodic(double battery, double[] currentDraw) {
    imuSim.addHeading(
        GetChassisSpeeds().omegaRadiansPerSecond
            * UnitConversion.DEGREES_PER_RADIAN
            * 20
            * UnitConversion.MILLISECONDS);

    frontLeft.SimPeriodic(battery, currentDraw);
    frontRight.SimPeriodic(battery, currentDraw);
    backLeft.SimPeriodic(battery, currentDraw);
    backRight.SimPeriodic(battery, currentDraw);
  }

  public ErrorCode SeedEncoders() {
    ErrorCode error = frontLeft.SeedTurnMotor();
    if (ErrorCode.OK != error) {
      return error;
    }

    error = frontRight.SeedTurnMotor();
    if (ErrorCode.OK != error) {
      return error;
    }

    error = backLeft.SeedTurnMotor();
    if (ErrorCode.OK != error) {
      return error;
    }

    return backRight.SeedTurnMotor();
  }

  public boolean Active() {
    return Math.abs(frontLeft.GetAngularVelocityRotationsPerSecond()) > 0.01
        || Math.abs(frontRight.GetAngularVelocityRotationsPerSecond()) > 0.01
        || Math.abs(backLeft.GetAngularVelocityRotationsPerSecond()) > 0.01
        || Math.abs(backRight.GetAngularVelocityRotationsPerSecond()) > 0.01;
  }

  public Pose2d GetPose() {
    return poseEstimator.getEstimatedPosition();
  }

  public boolean GetFieldCentric() {
    return fieldRelative;
  }

  public void SetFieldCentric(boolean fieldCentric) {
    fieldRelative = fieldCentric;
  }

  public ChassisSpeeds GetChassisSpeeds() {
    return measuredVelocity;
  }
}
